tidb: add explanation for the fields in `mysql.user` (#18318)
pingcap/docs: mysql-schema/mysql-schema-user.md

---
title: mysql_user
summary: Learn about the `user` table in the `mysql` schema.
---

# `mysql.user`

> `mysql.user` is a frequently used system table. You can display the columns of `mysql.user` by:

Reviewer: ```suggestion
The `mysql.user` table provides information about user accounts and their privileges.

To view the structure of `mysql.user`, use the following SQL statement:
```

```sql
DESC mysql.user;
```

> 

Reviewer: ```suggestion

The output is as follows:

```

| Reload_priv            | enum('N','Y')        | NO   |      | N                 |       |
| FILE_priv              | enum('N','Y')        | NO   |      | N                 |       |
| Config_priv            | enum('N','Y')        | NO   |      | N                 |       |
| Create_Tablespace_Priv | enum('N','Y')        | NO   |      | N                 |       |
| Password_reuse_history | smallint(5) unsigned | YES  |      | NULL              |       |
| Password_reuse_time    | smallint(5) unsigned | YES  |      | NULL              |       |
| User_attributes        | json                 | YES  |      | NULL              |       |
| Token_issuer           | varchar(255)         | YES  |      | NULL              |       |
| Password_expired       | enum('N','Y')        | NO   |      | N                 |       |
| Password_last_changed  | timestamp            | YES  |      | CURRENT_TIMESTAMP |       |
| Password_lifetime      | smallint(5) unsigned | YES  |      | NULL              |       |
+------------------------+----------------------+------+------+-------------------+-------+
44 rows in set (0.00 sec)
```

> There are several types of columns in `mysql.user`:

Reviewer: ```suggestion
The `mysql.user` table contains several fields that can be categorized into three groups:
```

<CustomContent platform="tidb">

* Scope:

>     * `Host` and `User` are used to specify a TiDB account

Reviewer: ```suggestion
    * `Host`: specifies the hostname of a TiDB account.
    * `User`: specifies the username of a TiDB account.
```

* Privilege:

>     * From `Select_priv` to `Drop_role_priv`, and from `Shutdown_priv` to `Create_Tablespace_Priv`: see [privileges required for TiDB operations](/privilege-management.md#privileges-required-for-tidb-operations)

Reviewer: ```suggestion

    The fields ending with `_priv` or `_Priv` define the permissions granted to a user account. For example, `Select_priv` means that the user has global `Select` privilege. For more information, see [Privileges required for TiDB operations](/privilege-management.md#privileges-required-for-tidb-operations).

```

* Security

>     * `authentication_string` and `plugin`: `authentication_string` records credentials for the accounts. Credentials are interpreted using the authentication plugin named in the `plugin` column.
>     * `Account_locked` records the account locking state.
>     * `Password_reuse_history` and `Password_reuse_time` are used for [password reuse policy](/password-management.md#password-reuse-policy)
>     * `User_attributes` provides information about user comments and user attributes
>     * `Token_issuer` is used for [`tidb_auth_token`](/security-compatibility-with-mysql.md#tidb_auth_token)
>     * `Password_expired`, `Password_last_changed` and `Password_lifetime` are used for [password expiration policy](/password-management.md#password-expiration-policy)

Reviewer: ```suggestion
    * `authentication_string` and `plugin`: `authentication_string` stores the credentials for the user account. The credentials are interpreted based on the authentication plugin specified in the `plugin` field.
    * `Account_locked`: indicates whether the user account is locked.
    * `Password_reuse_history` and `Password_reuse_time`: used for [Password reuse policy](/password-management.md#password-reuse-policy).
    * `User_attributes`: provides information about user comments and user attributes.
    * `Token_issuer`: used for the [`tidb_auth_token`](/security-compatibility-with-mysql.md#tidb_auth_token) authentication plugin.
    * `Password_expired`, `Password_last_changed`, and `Password_lifetime`: used for [Password expiration policy](/password-management.md#password-expiration-policy).
```

</CustomContent>

<CustomContent platform="tidb-cloud">

> * Scope:
>     * `Host` and `User` are used to specify a TiDB account
> * Privilege:
>     * From `Select_priv` to `Drop_role_priv`, and from `Shutdown_priv` to `Create_Tablespace_Priv`: see [privileges required for TiDB operations](https://docs.pingcap.com/tidb/stable/privilege-management#privileges-required-for-tidb-operations)
> * Security
>     * `authentication_string` and `plugin`: `authentication_string` records credentials for the accounts. Credentials are interpreted using the authentication plugin named in the `plugin` column.
>     * `Account_locked` records the account locking state.
>     * `Password_reuse_history` and `Password_reuse_time` are used for [password reuse policy](https://docs.pingcap.com/tidb/stable/password-management#password-reuse-policy)
>     * `User_attributes` provides information about user comments and user attributes
>     * `Token_issuer` is used for [`tidb_auth_token`](https://docs.pingcap.com/tidb/stable/security-compatibility-with-mysql#tidb_auth_token)
>     * `Password_expired`, `Password_last_changed` and `Password_lifetime` are used for [password expiration policy](https://docs.pingcap.com/tidb/stable/password-management#password-expiration-policy)

Reviewer: ```suggestion
* Scope:
    * `Host`: specifies the hostname of a TiDB account.
    * `User`: specifies the username of a TiDB account.
* Privilege:

    The fields ending with `_priv` or `_Priv` define the permissions granted to a user account. For example, `Select_priv` means that the user has global `Select` privilege. For more information, see [Privileges required for TiDB operations](https://docs.pingcap.com/tidb/stable/privilege-management#privileges-required-for-tidb-operations).

* Security
    * `authentication_string` and `plugin`: `authentication_string` stores the credentials for the user account. The credentials are interpreted based on the authentication plugin specified in the `plugin` field.
    * `Account_locked`: indicates whether the user account is locked.
    * `Password_reuse_history` and `Password_reuse_time`: used for [Password reuse policy](https://docs.pingcap.com/tidb/stable/password-management#password-reuse-policy).
    * `User_attributes`: provides information about user comments and user attributes.
    * `Token_issuer`: used for the [`tidb_auth_token`](https://docs.pingcap.com/tidb/stable/security-compatibility-with-mysql#tidb_auth_token) authentication plugin.
    * `Password_expired`, `Password_last_changed`, and `Password_lifetime`: used for [Password expiration policy](https://docs.pingcap.com/tidb/stable/password-management#password-expiration-policy).
```

</CustomContent>

> Most of the columns above exist in MySQL's `mysql.user`, except `Token_issuer`.

Reviewer: ```suggestion
While most of the fields in the TiDB `mysql.user` table are also present in the MySQL `mysql.user` table, the `Token_issuer` field is specific to TiDB.
```

* Security:
    * `authentication_string` and `plugin`: `authentication_string` stores the credentials for the user account. The credentials are interpreted based on the authentication plugin specified in the `plugin` field.
    * `Account_locked`: indicates whether the user account is locked.
    * `Password_reuse_history` and `Password_reuse_time`: used for [Password reuse policy](https://docs.pingcap.com/tidb/stable/password-management#password-reuse-policy).
    * `User_attributes`: provides information about user comments and user attributes.
    * `Token_issuer`: used for the [`tidb_auth_token`](https://docs.pingcap.com/tidb/stable/security-compatibility-with-mysql#tidb_auth_token) authentication plugin.
    * `Password_expired`, `Password_last_changed`, and `Password_lifetime`: used for [Password expiration policy](https://docs.pingcap.com/tidb/stable/password-management#password-expiration-policy).

</CustomContent>

> While most of the fields in the TiDB `mysql.user` table are also present in the MySQL `mysql.user` table, the `Token_issuer` field is specific to TiDB.

Reviewer 2: ```suggestion
Although most of the fields in the TiDB `mysql.user` table also exist in the MySQL `mysql.user` table, the `Token_issuer` field is specific to TiDB.
```

| Reload_priv            | enum('N','Y')        | NO   |      | N                 |       |
| FILE_priv              | enum('N','Y')        | NO   |      | N                 |       |
| Config_priv            | enum('N','Y')        | NO   |      | N                 |       |
| Create_Tablespace_Priv | enum('N','Y')        | NO   |      | N                 |       |
| Password_reuse_history | smallint(5) unsigned | YES  |      | NULL              |       |
| Password_reuse_time    | smallint(5) unsigned | YES  |      | NULL              |       |
| User_attributes        | json                 | YES  |      | NULL              |       |
| Token_issuer           | varchar(255)         | YES  |      | NULL              |       |
| Password_expired       | enum('N','Y')        | NO   |      | N                 |       |
| Password_last_changed  | timestamp            | YES  |      | CURRENT_TIMESTAMP |       |
| Password_lifetime      | smallint(5) unsigned | YES  |      | NULL              |       |
+------------------------+----------------------+------+------+-------------------+-------+
44 rows in set (0.00 sec)
```

There are several types of columns in `mysql.user`:

* Scope:

>   * `Host` and `User` are used to specify a TiDB account

Reviewer 3: ```
mysql-schema/mysql-schema-user.md: 69: MD007/ul-indent Unordered list indentation [Expected: 4; Actual: 2]
mysql-schema/mysql-schema-user.md: 71: MD007/ul-indent Unordered list indentation [Expected: 4; Actual: 2]
mysql-schema/mysql-schema-user.md: 73: MD007/ul-indent Unordered list indentation [Expected: 4; Actual: 2]
```

```suggestion
    * `Host` and `User` are used to specify a TiDB account
```

* Privilege:

>   * From `Select_priv` to `Drop_role_priv`, and from `Shutdown_priv` to `Create_Tablespace_Priv`: see [privileges required for TiDB operations](/privilege-management.md#privileges-required-for-tidb-operations)

Reviewer 3: ```suggestion
    * From `Select_priv` to `Drop_role_priv`, and from `Shutdown_priv` to `Create_Tablespace_Priv`: see [privileges required for TiDB operations](/privilege-management.md#privileges-required-for-tidb-operations)
```

* Security

>   * `authentication_string` and `plugin`: `authentication_string` records credentials for the accounts. Credentials are interpreted using the authentication plugin named in the `plugin` column.
>   * `Account_locked` records the account locking state.
>   * `Password_reuse_history` and `Password_reuse_time` are used for [password reuse policy](/password-management.md#password-reuse-policy)
>   * `User_attributes` provides information about user comments and user attributes
>   * `Token_issuer` is used for [`tidb_auth_token`](/security-compatibility-with-mysql.md#tidb_auth_token)
>   * `Password_expired`, `Password_last_changed` and `Password_lifetime` are used for [password expiration policy](/password-management.md#password-expiration-policy)

Reviewer 3: ```suggestion
    * `authentication_string` and `plugin`: `authentication_string` records credentials for the accounts. Credentials are interpreted using the authentication plugin named in the `plugin` column.
    * `Account_locked` records the account locking state.
    * `Password_reuse_history` and `Password_reuse_time` are used for [password reuse policy](/password-management.md#password-reuse-policy)
    * `User_attributes` provides information about user comments and user attributes
    * `Token_issuer` is used for [`tidb_auth_token`](/security-compatibility-with-mysql.md#tidb_auth_token)
    * `Password_expired`, `Password_last_changed` and `Password_lifetime` are used for [password expiration policy](/password-management.md#password-expiration-policy)
```
